Api v4 doc

ocket8888 opened the pull request:
Includes changes based on the WG discussion on Jan 14<sup>th</sup> 2020.

ocket8888 committed:
added an action

Set actions to run go tests

Uh oh, a bad test!

Fixed

Added tests for TO to actions workflow

Fixed Go version to our officially required 1.11

Also fixes treatment of inputs.

Added dependency fetch step

Added tests for all Go components

Fixed typo in package name

Added gofmt action

Added workflow for checking things when lib is updated

Introduced go fmt error in lib/

disabled main CI

Set specific types for PR events

Added TO workflow

Introduced compilation error into lib

fixed paths

fixed compilation error in lib

added build step to lib workflow

Fixed go fmt issues

Added weasel license checker

Fixed license issues

Removed unnecessary weasel action

Added docs build

Added docs error

(Hopefully) Fixed docs requirements for CI

Removed docs error

Added runtime dependency for docs CI

Installing Python3 version of PSUtil

Fixed clearing of sphinxopts in Makefile

ocket8888 committed:
Added APIv3 document thingo

ocket8888 committed:
Added proper type/representation of empty collection properties

ocket8888 committed:
Added note about request path suffixes

ocket8888 committed:
Added definition of tenantable objects

ocket8888 committed:
Added build artifacts to gitignore

ocket8888 committed:
Added admonition about no duplicate endpoint functionality

Also relaxed restriction on path parameters. I realize those aren't totally compatible
but a discussion about that is upcoming.

ocket8888 committed:
Added abstract, clarified null-vs-zero values for primitives

Also switched the default output timestamp format to RFC3339 and removed the bit about
optional support of multiple definitions of 'sort_by'.

ocket8888 committed:
Added authentication section

ocket8888 committed:
Added physical locations and tags to data model

ocket8888 committed:
Added section on user permissions

ocket8888 committed:
Added a bunch of path-parameter endpoints

ocket8888 committed:
Incremented version

jhg03a commented on blueprints/APIv3.tex:
```diff
@@ -461,7 +478,79 @@ \subsection{Relationships as Objects}
+scheme.\\
+
+\subsection{JWT Authentication}
+Alternatively, authentication can be done using OAuth2 Bearer tokens in JSON Web-Token (JWT) format\footnote{The intention of this section
```
Might be good to specify which RFC apply to these JWT.  There are several for separate purposes.

ocket8888 replied:
I did specify a couple later on - was it not a complete list/do you think I should move them to earlier in the section?

jhg03a replied:
I'd say just move it earlier.

ocket8888 replied:
It actually looks like I was missing number 7519, but that should be fixed now. Better?

jhg03a replied:
Yep

jhg03a commented on blueprints/APIv3.tex:
```diff
@@ -461,7 +478,79 @@ \subsection{Relationships as Objects}
+
+\begin{itemize}
+	\item \code{iss} The issuer of the JWT. By default this should have the value of the network location of the Traffic Ops server that
+	generated the JWT - e.g. "trafficops.infra.ciab.test" - but may also refer to the LDAP server used when users authenticate using LDAP.
```
The issuer field should always refer to the system which generated the JWT.  You can't get a JWT directly from ldap, so it should refer to the intermediary system that signed the JWT.

jhg03a commented on blueprints/APIv3.tex:
```diff
@@ -461,7 +478,79 @@ \subsection{Relationships as Objects}
+	generated the JWT - e.g. "trafficops.infra.ciab.test" - but may also refer to the LDAP server used when users authenticate using LDAP.
+	\item \code{sub} The "subject" of the JWT - in this case, the Traffic Ops user. This MUST be a \emph{full} representation of the
+	authenticated user according to Section \ref{sec:users}, and must be serialized as a JSON object.
+	\item \code{exp} The timestamp at which this JWT will expire. This MUST coincide with the expiration date/time of the mojolicious
```
JWT issuers aren't required to set expiration in a particular way.  If you're using the bearer token to basically trade in for a mojo cookie, it should be evaluated at the time of the trade-in for validity.

ocket8888 replied:
The RFC says it has to be a Unix epoch timestamp in seconds - or were you talking about how it "MUST coincide with the expiration date/time of the mojolicious cookie"? The intent there is that you can't authenticate after your authorization expires, but TO sets the cookie on every request, thereby resetting the expiration. Idk of any mechanism that would do that with a token, so it may not be appropriate to require that coincidence.

jhg03a replied:
The coincidence was my main objection there.  If the JWT is valid, it shouldn't matter how that translates to mojo cookies.

jhg03a replied:
You implicitly get the ability to blacklist a JWT by changing the role of a user.

jhg03a commented on blueprints/APIv3.tex:
```diff
@@ -492,9 +585,9 @@ \subsection{Traffic Stats Servers}
+\subsection{Users\label{sec:users}}
 
-\section{Tenantable Objects}
+\section{Tenantable Objects\label{sec:tenantable}}
```
I'd also say there should be some objects which are specifically belonging to the root tenant including things such as cachegroups, asns, physical locations, servers, etc for security purposes.

ocket8888 replied:
Those things don't belong to any tenant today - they aren't tenantable. Are you saying that If I'm on any child tenant then I shouldn't be allowed to see cachegroups, asns, physical locations or servers at all?

jhg03a replied:
Correct.  I view those pieces of information as privileged because they could be used to identify weak points in the overall architecture or contain contact information for facilities to be used in spearphishing attacks.

jhg03a replied:
I'd be ok if it were just a separate role capability too rather than requiring the root tenant so it could be granted on exception.  That's actually probably a better answer than requiring the root tenant because the root tenant exposes all plaintext configuration on a DS.

ocket8888 replied:
Also, I'm not on our root tenant, so that'd leave me pretty crippled :P.

Yeah, I don't think Tenancy should be used to limit what _kind_ of objects someone can see/manipulate, only _which_ objects. That's what User Permissions are for.

ocket8888 committed:
Fixed incorrect constraint on 'iss' field

mitchell852 commented:
I'd love to see support for searching. I.e. the following query params or something along these lines:

```
&search[value]=foo
&search[columns]=a,b,c,d (to limit the search to certain columns if desired)
&search[regex]=true|false (or maybe just let the search value support globbing?)
```

and because it's important to know how many records there are before a search (or any filter for that matter) was applied maybe a response like this:

```
{
	response: []
	alerts: []
	meta: { total: 200, otherfieldTBD: ..., otherfieldTBD: ... }
}
```

ocket8888 committed:
Fixed grammatical error

ocket8888 committed:
Added missing JWT RFC

ocket8888 committed:
Removed cookie-and-token-expiry coincidence clause

ocket8888 committed:
Split sections into their own files

ocket8888 commented:
> _"I'd love to see support for searching. I.e. the following query params or something along these lines:"_
> _```&search[value]=foo
&search[columns]=a,b,c,d (to limit the search to certain columns if desired)
&search[regex]=true|false (or maybe just let the search value support globbing?)```_

That seems extremely complicated. First, I'd vote no regex, that's unlikely to be helpful and clients should largely be in charge of such complex filtering IMO. Having a single value that applies to arbitrary "columns" not only ties the representations directly to table columns in the database - which won't necessarily be true - but also sounds like a nightmare to implement. I also hope this doesn't turn into something that people turn into something like what "Types" are now; e.g. "For this part of Traffic Router, select all the Cache Servers that have a Profile name that matches `Foo*` - because those are Foo Edge Cache Servers. Tags should ideally fill that role if necessary, IMO.

If you really think this is necessary - I'm not still quite convinced of clients' performance-driven need for that - then what I'd suggest is: object property filtering query parameters support the use of `*` for globbing in the case of string fields. So if you want all the Cache Servers that have a Profile name that matches `Foo*` you'd request `GET /api/3.0/cache_servers?profile=Foo* HTTP/1.1`. This obviously won't work on numeric fields like e.g. "ID" but somewhat less obviously also won't work on timestamps - or at least shouldn't, because that's what Age Filtering is for.

ocket8888 commented:
> _"... because it's important to know how many records there are before a search (or any filter for that matter) was applied maybe a response like this"_

This 'meta' field for API responses does seem useful, even outside the context of server-side searches. I just have trouble deciding how it should be standardized. It's important because of statistics endpoints, they would ideally use it as well to provide a more consistent form, but I'll need to spend some time digging into all the statistics currently available through the API and what kinds of information they return, what properly belongs in `response` and what might use "meta". Also kinda want to try to workshop that name.

ocket8888 committed:
Added a bit to the data model about DSes (active vs primed)

ocket8888 committed:
Laid out data types and fallbacks

ocket8888 committed:
Added CDN data model

ocket8888 committed:
Added Cache Server section content

jhg03a commented:
It seems like we're starting to rewrite our help documentation inside the api documentation.

ocket8888 committed:
Added User data model section

Also switched Delivery Services from having Primed/Active as separate, boolean properties to
using one 'Status' property that encapsulates both concepts.

ocket8888 commented:
> _"It seems like we're starting to rewrite our help documentation inside the api documentation."_

What do you mean?

ocket8888 committed:
Fleshed out DS property list, still missing some things

ocket8888 committed:
formatting

ocket8888 committed:
Added Tenant property description

ocket8888 committed:
Added supported protocols and bypass destination

ocket8888 committed:
Finished bypass destination description

jhg03a commented on blueprints/datamodel.tex:
```diff
@@ -335,7 +335,117 @@ \subsection{Delivery Services}
+actually be done is dependent on the configuration of each Traffic Router itself,
+and if this Delivery Service is DNS-routed the only IP address Traffic Routers
+will be capable of checking for anonymization (e.g. known proxy/VPN/TOR exit node)
+then
```
Note this is a global configuration underneath this boolean flag.  There's an open issue to allow two DS to vary in the configuration.

ocket8888 replied:
So you're saying that anonymous blocking is configured on a Traffic Router today, and individual DS settings have no effect?

jhg03a replied:
You can't express different settings per DS, it's all or nothing per ds and one config per cdn.  https://github.com/apache/trafficcontrol/issues/4227

ocket8888 replied:
Well you can't define subsets of anonymous blocking or IP whitelists on a per-Delivery Service basis, but as I understand that issue the simple "check for anonymization vs don't" flag does work as advertised. Correct?

Do you think Delivery Services also need a property to define IP whitelists for the purposes of anonymous blocking and/or geographic blocking? And maybe also a "policy" of what to consider "anonymized"? That second one sounds pretty hard to do without digging into TR implementation details.

jhg03a replied:
That's not how I've understood it.  Per the docs it's a TR profile setting which per CDN, and it's only a boolean to turn whatever that profile has enabled on or off.

jhg03a commented on blueprints/datamodel.tex:
```diff
@@ -335,7 +335,117 @@ \subsection{Delivery Services}
+
+\paragraph{Deep Caching}
+
+\paragraph{Caching}
```
Cachability might be a better word here.

ocket8888 replied:
I'm a little hesitant to use made-up words. Is there any other phrase you think might be more appropriate? Like "Content Caching" or "Caching Type"?

jhg03a commented on blueprints/datamodel.tex:
```diff
@@ -335,7 +335,117 @@ \subsection{Delivery Services}
+%TODO: MSO use-case?
+\paragraph{Origin}
+
+\paragraph{Query String Handling}
```
I believe we should kill this option all together and force people to actually think about their cachekey.  Most HTTP security classes that mention CDNs talk specifically about this concept.

ocket8888 replied:
Well how would that be implemented today using ATS configuration files? I was under the impression that cache key configuration required a plugin that was totally separate from remap.config

jhg03a replied:
We do it today using a combination of raw remap and DS profile/parameters.  This is required today for every base origin url field change or you end up breaking existing caches.

ocket8888 replied:
Wait, so every time you change the Delivery Service's orgServerFqdn you have to change the cache key using raw remap.config lines? Why? And what Parameters?

jhg03a replied:
Correct.  If you don't manually assert the cachekey config before changing the base origin url field, you effectively invalidate the entire cache of that ds.

jhg03a commented on blueprints/datamodel.tex:
```diff
@@ -335,7 +335,117 @@ \subsection{Delivery Services}
+
+\paragraph{Range Request Handling}
+\begin{itemize}
+	\item \code{NO\_CACHE}
```
These list items don't match how our existing range request options work.

ocket8888 replied:
It matches [Range Request Handling](https://traffic-control-cdn.readthedocs.io/en/latest/overview/delivery_services.html#range-request-handling) on current Delivery Services.

jhg03a replied:
Not quite, and there are more options now.  We're just hand crafting them into other fields because the field doesn't support them.

ocket8888 replied:
What other fields?

jhg03a replied:
Header rewrites and the raw remap

ocket8888 replied:
In what ways? I'd like to expand this property to encompass the different handling methods.

jhg03a commented on blueprints/datamodel.tex:
```diff
@@ -335,7 +335,117 @@ \subsection{Delivery Services}
+	\item \code{HTTP}
+	\item \code{DNS}
+	\item \code{STEERING}
+	\item \code{STATIC}
```
By static, I assume you mean ANY_MAP.  This is not an option anyone without operator should know about because it's capable of harming traffic outside a tenant.

ocket8888 replied:
Well no, actually I mean Federations. My hope is that ANY_MAP can be done away with entirely, but that's still something to be discussed.

Since the whole code base is open source, though, I see  little point in not documenting things just because a non-operator can do harm with them. In 2.2 I can count at least three ways to gain root access on a machine running TO - that wouldn't be safer if they weren't documented. That ought to just be not possible.

jhg03a replied:
ANY_MAP is unfortunately still required in dire circumstances.  But a general user shouldn't know about it.

ocket8888 replied:
Okay, but this is a full data model definition, not a help guide for general user.

Like I said, we'll still need to discuss ANY_MAP because I don't actually know what "dire circumstances" necessitate it.

jhg03a commented on blueprints/datamodel.tex:
```diff
@@ -335,7 +335,117 @@ \subsection{Delivery Services}
+\begin{itemize}
+	\item \code{HTTP}
+	\item \code{DNS}
+	\item \code{STEERING}
```
Are we finally killing steering off and going to merge its use cases with Client Steering?

ocket8888 replied:
Yes.

jhg03a replied:
What's the plan for steering regex then?

ocket8888 replied:
Depends: what's that?

Absolutely nothing in any section above 2 (and 2 only as of this past Tuesday) has been discussed. They should be considered **very** immature works in progress that can and will change without warning as I change my mind and learn more about specific use-cases etc.

jhg03a commented on blueprints/datamodel.tex:
```diff
@@ -355,10 +465,33 @@ \subsubsection{Status}
 
+\paragraph{Tenant}
+
+\paragraph{Vanity Hostnames}
```
If this is to replace HOST_REGEXs, what do you intend to do about the other kinds of regexes?

ocket8888 replied:
Nothing, yet. Still has to be discussed because I don't understand the full use-case set for custom match lists.

jhg03a commented on blueprints/datamodel.tex:
```diff
@@ -355,10 +465,33 @@ \subsubsection{Status}
+
+\paragraph{Logged Request Headers}
+
+\paragraph{Significant Query Parameters}
```
See above comment about promoting the cachekey concept.

ocket8888 replied:
This doesn't have anything to do with caching, it's the "Consistent Hashing Query Parameters". Maybe the property needs a different name, but the only other one I could come up with was "Query Parameters Significant for Routing" which is a pretty long name.

jhg03a replied:
Or think about it as overlapping subsets of cachekey config that spans both ATS and TR.  You'd never configure TR to do one thing while ATS to do another (assuming you weren't off in the weeds of ATS power).

ocket8888 replied:
Yeah, one would hope those line up.

jhg03a commented:
I mostly comment about feeling like the API doc scope is creeping because we're including redefinitions/copypasta from our existing documentation.  If we intend to keep the same fields and same concepts, we might want to reference our existing help documents.  This would call out places that are specifically different, or allows for an explanation as to what happened.

jhg03a commented on blueprints/datamodel.tex:
```diff
@@ -335,7 +335,117 @@ \subsection{Delivery Services}
+
+%TODO: use-case for header-rewrite text?
+
+\paragraph{Miss Location}
```
This option should not be available to anyone without operator rights.  It's never correct to allow average users to specify what the default geo-miss location is or they could crush arbitrary cachegroups possibly.

ocket8888 replied:
"Operator rights" will ideally be expressed as a set of user permissions that fall within the "operator" role. Is there a specific permission you have in mind, or do "permissions" not properly encapsulate the permission you are trying to express?

jhg03a replied:
Today the permission model applies to tables for the most part.  This field shouldn't be exposed to anyone without high permissions because of the potential for abuse.

ocket8888 replied:
So you're saying that, in addition to the `deliveryservices-write` Permission someone  should also need the e.g. `deliveryservices-misslocation` permission to set the Miss Location of a Delivery Service to anything non-null?

jhg03a replied:
Yes

jhg03a commented:
Who is the anticipated audience for this document?  Does this entail just patching the API, or adjusting both the API and the underlying postgres datamodel?

ocket8888 commented:
> _"I mostly comment about feeling like the API doc scope is creeping because we're including redefinitions/copypasta from our existing documentation. If we intend to keep the same fields and same concepts, we might want to reference our existing help documents. This would call out places that are specifically different, or allows for an explanation as to what happened."_

Yes and no. Existing documentation would be superseded and replaced by this documentation, but in most cases the semantic meaning of objects have changed. I don't mean for this to be a full re-telling of ATC concepts (e.g. the "Deep Caching" property of a Delivery Service is described as defining "whether or not 'Deep Caching' is permitted to be used by the Delivery Service") but a full definition of its scope is necessary because many scopes have changed.

ocket8888 commented:
> _"Who is the anticipated audience for this document?"_ 

The intended audience is primarily TO API authors, but also TO API consumers - including other ATC component developers - and particularly those who develop user-land clients for the TO API.

> _"Does this entail just patching the API, or adjusting both the API and the underlying postgres datamodel?"_

This defines a data model to which both the API and database must adhere. They need not implement things the same way, but must ultimately express the same concepts herein defined as the data model.

ocket8888 committed:
Made JWT 'sub' field spec smaller

Also some formatting changes

ocket8888 committed:
Removed objects from CDN model that don't need to be in a CDN

ocket8888 committed:
Added descriptions for 'Caching' DS property

ocket8888 committed:
Added DSCP description

jhg03a commented:
One other tidbit to throw in, in addition to the markdown version of this document I'd also like to see a programatic schema in either [JSON Schema](http://json-schema.org/) or [Swagger / Open API Spec](https://swagger.io/).  I believe that would help with API stability and testability significantly.

ocket8888 commented:
You know, I had thought about including Typescript interface definitions for the data model, but idk how many people know typescript - or even care since none of our main components use that. The JSON schema thing (which OAS uses too) looks kind of interesting, if verbose. The only thing that concerns me is that given the schema:
```json
{
	"type": "object",
	"title": "SomeType",
	"properties": {
		"foo": {"type": "string"},
		"test": {"type": "string"}
	},
	"required": ["foo"]
}
```
... this is for some reason considered totally valid:
```json
{
	"foo": "bar",
	"otherProperty": {}
}
```
... which is totally strange to me. It doesn't seem to have a way - that I can see - to be less permissive. Also seems kinda verbose to express things like "properties that must exist". But if that's the best way to strictly define the model in a machine-readable way, I can probably make do. I'm not going to even think about doing that until the data model is actually defined, though.

For the record, the equivalent in Typescript is simply:
```typescript
interface SomeType {
	foo: string;
	test?: string;
}
```

edit: the Typescript interface *kind of* has the same problem as JSON Schema w/r to permissiveness. Given the above interface definition, you technically can do e.g. `const st = {foo: "bar", otherProperty: {}} as SomeType;`, but this sort of casts the extraneous fields aside so that later doing e.g. `let a = st.otherProperty;` is a compile error. And you still can't do `const st: SomeType = {foo: "bar", otherProperty: {}};`. But that's all pretty implementation-dependent. Which is maybe a problem.

ocket8888 committed:
Added Vanity Hostnames description

ocket8888 committed:
Removed DNSSEC Keys from CDN model

ocket8888 committed:
Added requirement for 'summary.count' for paginated results

ocket8888 committed:
Removed reserved summary properties 'min', 'max', and 'mean'

These are now mandated names for the concepts, but do not reside in the 'summary'.

ocket8888 committed:
PUT requires If-Not-Modified-Since, PATCH requires ETag/If-Match

ocket8888 committed:
Added method for obtaining status of async tasks

ocket8888 committed:
Added requirement to reject payloads with extra keys

ocket8888 committed:
Restricted 403 responses to not disclose why access was forbidden

ocket8888 committed:
Removed page and page_size query parameters

ocket8888 committed:
Added restrictions on identifier values

ocket8888 committed:
Added section on ETag generation

ocket8888 committed:
Formatting changes

ocket8888 committed:
Removed statement about using 404 for identifying property filters

ocket8888 committed:
Fixed typo

ocket8888 committed:
Updated asynchronous status update endpoints to include start/end times and exclude 'waiting' status

ocket8888 committed:
Added requirement for error-level alert in responses encoding FAILED jobs

ocket8888 committed:
Specified default sort-order ('ascending')

ocket8888 committed:
Made it very clear what happens at edge cases in age filtering

ocket8888 committed:
Changed Tenancy behavior requirements to be 'omit from collections, 403 on direct request')

ocket8888 committed:
Changed query parameters from snake_case OR camelCase to ONLY camelCase

ocket8888 committed:
Formatting fixes

ocket8888 committed:
Fix typo

ocket8888 committed:
Added TS data model, added Cache Server and TM data defs

ocket8888 committed:
Fixed a typo

ocket8888 committed:
Clarified connection between hostname and domain

ocket8888 committed:
HTTP and HTTPS ports are not allowed to be null

ocket8888 committed:
loadavg moved to Profile, can be 0, and only counts for the one-minute value

ocket8888 committed:
Moved monitoring thresholds and storage config to Profile level

ocket8888 committed:
Changed the 'no limit' semantic value from 0 to null in several places

ocket8888 committed:
Added Cache Group definitions

ocket8888 committed:
Added Cache Group Type

ocket8888 committed:
Got rid of 'skipMidTier'

ocket8888 committed:
Allowed maxOriginConnections to be null

ocket8888 committed:
Added Tags to Profiles and Role descriptions

ocket8888 committed:
Finished Profiles section

ocket8888 committed:
Added Tags to Cache Groups and Delivery Services

ocket8888 committed:
Added disclaimer to Vanity Hostnames

ocket8888 committed:
fixed typo in footnote

ocket8888 committed:
Added tags to users

ocket8888 committed:
Added license headers to ts datamodel

ocket8888 committed:
Created subdirectory for datamodel parts

ocket8888 committed:
Broke the data model out into separate files

ocket8888 committed:
Added definiton of GET method of /cdns

ocket8888 committed:
Added type and cdn to profiles

ocket8888 committed:
Added spec for GET /profiles

ocket8888 committed:
Started adding POST /profiles spec

ocket8888 committed:
Added more profile route definitions

ocket8888 committed:
Changed APIv3 -> API v4

ocket8888 committed:
Added workflow for building RPMs

ocket8888 committed:
fixed docker install command

ocket8888 committed:
install lsb-release package

ocket8888 committed:
call lsb_release instead of lsb-release

ocket8888 committed:
Switch to official docker image

ocket8888 committed:
Make entrypoint executable

ocket8888 committed:
Need to install bash

ocket8888 committed:
Fix bad interpreter used for entrypoint

ocket8888 committed:
Add debug output

ocket8888 committed:
bind stderr to stdout; remove debug output

ocket8888 committed:
There's now no way for output to escape

ocket8888 committed:
Force setting -ex in pkg

ocket8888 committed:
explicit project list

ocket8888 committed:
Fixed typo

ocket8888 committed:
Attempt to fix path to docker-compose file

ocket8888 committed:
Attempt to fix my problems by manually installing docker-compose

ocket8888 committed:
cleanup

ocket8888 committed:
Added CDN-in-a-Box action and workflow step

ocket8888 committed:
Switch from using 'pkg' to direct rpm builds

ocket8888 committed:
Switched container from docker to go

ocket8888 committed:
Added git

ocket8888 committed:
go doesn't support GNU-standard args parsing..............

ocket8888 committed:
added go to PATH

ocket8888 committed:
Added some GOPATH trickery

ocket8888 committed:
Added go get

ocket8888 committed:
Added license header

ocket8888 committed:
Prove that the thing can be built

ocket8888 committed:
try building without links

ocket8888 committed:
set -ex

ocket8888 committed:
added rsync

zrhoffman committed:
Wrap docker command to build Traffic Ops in JavaScript action

ocket8888 committed:
Added definition of Infrastructure Servers

ocket8888 committed:
Added spec for DELETE /profiles/{{name}}/parameters

ocket8888 committed:
Added definition of Physical Locations

ocket8888 committed:
Added definition of a Traffic Portal

ocket8888 committed:
Added definition of a Tenant

ocket8888 committed:
Added definition of Capabilities

ocket8888 committed:
Removed Implementation Details

ocket8888 committed:
Incremented version

ocket8888 committed:
Slight refactor and now builds all RPMs

ocket8888 committed:
Fixed a typo

ocket8888 committed:
Added license header

ocket8888 committed:
Added timing info to logs

ocket8888 committed:
Removed bash-isms

ocket8888 committed:
Added git

ocket8888 committed:
Specify only the necessary containers to build and start

ocket8888 committed:
Cleanup and renaming

ocket8888 committed:
Added running the integration tests

ocket8888 committed:
Added logs on failure

ocket8888 committed:
Using absolute paths to mounted scripts

ocket8888 committed:
Added Tags to physical locations

ocket8888 committed:
Added note about miss locations to DS model

ocket8888 committed:
Added definition of an Origin

ocket8888 committed:
Version increment

ocket8888 committed:
Added definition of a Traffic Vault

ocket8888 committed:
Added coordinates to origins, removed physical locations

ocket8888 committed:
Added health polling format to profiles

ocket8888 committed:
Added history count to profiles

ocket8888 committed:
Added a bunch of TR stuff

ocket8888 committed:
Version bump

ocket8888 committed:
Finished modeling Traffic Router

ocket8888 committed:
Added description of Topologies

ocket8888 committed:
Added beginnings of Traffic Stats Server definition

ocket8888 committed:
Fixed typo

ocket8888 committed:
Removed TS-to-CDN associations; removed graph URLs from TS servers

ocket8888 committed:
Added database servers to Traffic Stats servers

ocket8888 committed:
Made Cache Maitenance Interval nullable

ocket8888 committed:
Added HTTPS port to TR instances

ocket8888 committed:
Fixed incorrectly referencing Cache Server in TR spec

ocket8888 committed:
Added notes to vaults

ocket8888 committed:
Added 'online' property to all non-cache servers

ocket8888 committed:
Added health connection timeout to Profiles

ocket8888 committed:
Added definition of global configuration object

ocket8888 committed:
Added definition of Invalidation Job objects

ocket8888 committed:
Added note about TOAPI proxying to TR Online property

ocket8888 committed:
health connection timeout no longer allowed to be zero

ocket8888 committed:
Added Traffic Ops server definition

ocket8888 committed:
Added endpoint definition for /tags

ocket8888 committed:
Added endpoint spec for /tags/{{Tag Name}}

ocket8888 committed:
Relaxed name restrictions on Tags

ocket8888 committed:
Moved TypeScript interfaces into actual document

ocket8888 committed:
Moved Traffic Router interface into document

ocket8888 committed:
Minor DS work

ocket8888 committed:
Fixed unescaped '_'

ocket8888 committed:
fix incorrect example in API endpoint spec

ocket8888 committed:
Added tag name length limit

ocket8888 committed:
Maade user email non-null

ocket8888 committed:
Fixed broken Typescript for Traffic Monitors

ocket8888 committed:
Add minimum pre-emptive start time to Invalidation Jobs

ocket8888 committed:
Invalidation Jobs are now tenant-able

ocket8888 committed:
Model capabilities in TS

ocket8888 committed:
Model profiles in TS

ocket8888 committed:
Move DS properties to appropriate subtypes

Also adds descriptions for many previously un-described properties, and adds Typescript definitions for all non-MSO sub-types

ocket8888 committed:
Remove now-unneccessary footnote

ocket8888 committed:
Fix some typescript typos, extract target type definition

ocket8888 committed:
Added MSO DS definitions

ocket8888 committed:
Added definition of Steering-Routed DS properties

ocket8888 committed:
Fix rendering of toc with \code in section headings

ocket8888 committed:
Add missing period

ocket8888 committed:
Add endpoint spec for /traffic_portals

ocket8888 committed:
Added spec for POST handler of /traffic_portals

ocket8888 committed:
Added spec for /traffic_portals/{{URL}}

ocket8888 committed:
Fixed missing commas in example JSON

ocket8888 committed:
Fixed broken table reference

ocket8888 committed:
Added note about returning a warning when lookup fails
